shithead_game_server/src/lobby/in_game.rs

```diff
@@ -0,0 +1,160 @@
+        let _ = self
+            .lobby_data
+            .broadcast_messages_sender
+            .send(ServerMessage::Turn(new_turn_player_id));
+    }
+
+    /// The current turn has timed out, give the current turn player the trash or play one of his
+    /// cards, and advance to the next turn.
+    pub async fn turn_timeout(&mut self) {
+        // give the player all the cards in the trash
+        self.lobby_data.player_list.give_cards_to_player(
+            self.in_game_state.current_turn.player_id(),
+            self.in_game_state.trash.take_all(),
+        );
+
+        // let all the players in the lobby know that the player who timed out got all cards from
+        // the trash.
+        let _ = self
+            .lobby_data
+            .broadcast_messages_sender
+            .send(ServerMessage::GiveTrash(
+                self.in_game_state.current_turn.player_id(),
+            ));
+
+        self.advance_turn_and_update_players().await
+    }
+
+    /// The player has finished playing his turn, move on the next turn.
+    pub async fn turn_finished(&mut self) {
+        self.advance_turn_and_update_players().await;
+    }
+
+    /// Handles a card click from one of the clients.
+    pub async fn click_card(
+        &mut self,
+        client_id: ClientId,
+        clicked_card_location: ClickedCardLocation,
+    ) -> Result<(), GameServerError> {
+        let current_turn_player_id = self.in_game_state.current_turn.player_id();
+
+        if current_turn_player_id != client_id {
+            return Err(GameServerError::NotYourTurn);
+        }
+
+        let player = self
+            .lobby_data
+            .player_list
+            .get_player(client_id)
+            .ok_or(GameServerError::NotInALobby)?;
+
+        match clicked_card_location {
+            ClickedCardLocation::Trash => {
+                // if the player can place any of his cards in the trash, then he is not allowed to
+                // take the trash
+                if player
+                    .what_cards_can_be_placed_on(self.in_game_state.trash_top_card_rank())
+                    .next()
+                    .is_some()
+                {
+                    return Err(GameServerError::CantTakeTrashBecauseSomeCardsCanBePlayed);
+                }
+
+                // the player has finished his turn
+                self.turn_finished().await;
+
+                // give the player all the cards in the trash
+                self.lobby_data.player_list.give_cards_to_player(
+                    current_turn_player_id,
+                    self.in_game_state.trash.take_all(),
+                );
+
+                // let all the players in the lobby know that the player who timed out got all cards from
+                // the trash.
+                let _ = self
+                    .lobby_data
+                    .broadcast_messages_sender
+                    .send(ServerMessage::GiveTrash(current_turn_player_id));
```

Comment:
The `turn_timeout()` method does what's written in here. Also, I think that every `Turn` struct creates a timeout that calls the `turn_timeout()` method, so maybe this implementation will be executed twice.

Reply:
In the future the `turn_timeout` will play one of the player's cards if possible, so its behaviour is not exactly the same.